cmd/thanos/check.go

```diff
@@ -0,0 +1,119 @@
+package main
+
+import (
+	"io/ioutil"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	thanosrule "github.com/improbable-eng/thanos/pkg/rule"
+	"github.com/oklog/run"
+	"github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/pkg/rulefmt"
+	"github.com/prometheus/tsdb"
+	"gopkg.in/alecthomas/kingpin.v2"
+	"gopkg.in/yaml.v2"
+)
+
+func registerChecks(m map[string]setupFunc, app *kingpin.Application, name string) {
+	cmd := app.Command(name, "Linting tools for Thanos")
+	registerCheckRules(m, cmd, name)
+}
+
+func registerCheckRules(m map[string]setupFunc, root *kingpin.CmdClause, name string) {
+	checkRulesCmd := root.Command("rules", "Check if the rule files are valid or not.")
+	ruleFiles := checkRulesCmd.Arg(
+		"rule-files",
+		"The rule files to check.",
+	).Required().ExistingFiles()
+
+	m[name+" rules"] = func(g *run.Group, logger log.Logger, reg *prometheus.Registry, _ opentracing.Tracer, _ bool) error {
+		// Dummy actor to immediately kill the group after the run function returns.
+		g.Add(func() error { return nil }, func(error) {})
+		return checkRulesFiles(logger, ruleFiles)
+	}
+}
+
+func checkRulesFiles(logger log.Logger, files *[]string) error {
+	failed := tsdb.MultiError{}
+
+	for _, f := range *files {
+		n, errs := checkRules(logger, f)
+		if errs.Err() != nil {
+			level.Error(logger).Log("result", "FAILED")
+			for _, e := range errs {
+				level.Error(logger).Log("error", e.Error())
+				failed.Add(e)
+			}
+			level.Info(logger).Log()
+			continue
+		}
+		level.Info(logger).Log("result", "SUCCESS", "rules found", n)
+	}
+	if failed.Err() != nil {
+		return failed
+	}
+	return nil
+}
+
+func checkRules(logger log.Logger, filename string) (int, tsdb.MultiError) {
+	level.Info(logger).Log("msg", "checking", "filename", filename)
+	checkErrors := tsdb.MultiError{}
+
+	b, err := ioutil.ReadFile(filename)
+	if err != nil {
+		checkErrors.Add(err)
+		return 0, checkErrors
+	}
+
+	var rgs thanosrule.RuleGroups
+	if err := yaml.Unmarshal(b, &rgs); err != nil {
```

Comment:
One more small thing to think about: maybe it would be worth to be strict here too like in `promtool` and use `UnmarshalStrict`? That way we would ensure that no unknown fields are used.

Reply:
@FUSAKLA @bwplotka @povilasv what is your opinion on this?

Reply:
Hm, the thing is Thanos rule uses also Unmarshal only. 
Promtool uses strict because Prometheus uses strict. 

Strict can help you with typos for example instead using some defaults and ignoring the misspelled field but I also like the consistency. 

Maybe in next PR make it strict in both cases and use the validation even on the Thanos rule start?

Reply:
IMO, let's do strict here and do a PR to change Thanos RULE to use strict. Nice catch :100:

Reply:
Ok sounds good to me.
Will switch it to the strict mode.

Reply:
Taking it back.. currently this cannot be done because the unmarshaling is done in two steps. 
First unmarshals without strict only the struct with partial response and in second step unmarshals the imported prometheus rule group also with unstrict unmarshall so it passes the additional partial_response_strategy field.

I'll take a closer look but now it fails with the strict.

```go
rs := struct {
		String string `yaml:"partial_response_strategy"`
	}{}

	errMsg := fmt.Sprintf("failed to unmarshal 'partial_response_strategy'. Possible values are %s", strings.Join(storepb.PartialResponseStrategyValues, ","))
	if err := unmarshal(&rs); err != nil {
		return errors.Wrapf(err, errMsg)
	}

	rg := rulefmt.RuleGroup{}
	if err := unmarshal(&rg); err != nil {
		return errors.Wrapf(err, errMsg)
	}
```

Reply:
Ah I see. Maybe we can add our own RuleGroup struct which embedds prometheus one?

```
type RuleGroup struct{
    PartialResponseStrategy string `yaml:"partial_response_strategy"`
    rulefmt.RuleGroup
}
```

This would simplify code, and then we could do unmarshal in one pass + use strict. WDYT?

Reply:
Uh.. looking into it more it's not that simple. `PartialResponseStrategy` is defined in protobuffer and is `int32` the `RuleGroup` defined in `rule.go` has overridden the `UnmarshalYAML` where the mapping from `WARN` to `0` is done. Also setting the default value if the `partial_response_strategy` is missing at all. 

Not sure if it is possible to add own `UnmarshalYAML` to the protobuf type? This could possibly simplify it? 

Maybe @bwplotka could add more info to this I remember we discussed this a bit when he was adding the partial response strategy.

Reply:
Sorry, I didn't notice you changed the type to the `string`. 
This way it can be done but there will be two new structs instead using the Thanos rule ones.

Reply:
Is this what you meant? https://github.com/improbable-eng/thanos/pull/1097/commits/864a74cafb1f301143f11bf5c88cd6333ce1413b

TBH I don't like making the lint and actual loading rules in rule more different but if we eventually unify it in some next PR than I'm OK with it I suppose.

Reply:
Yes, thats what I meant but I think you could do `var rgs []ThanosRuleGroup` instead of `var rgs ThanosRuleGroups`, no ? so then it's only 1 struct :)

Reply:
Great, I thing this cannot be done because I need to unmarshal  the `groups: ` key to the slice.
I can't unmarshal it to slice directly. Or maybe I misunderstood?

Reply:
makes sense. all good :+1: